pkg/trace/api: add generic EVP Proxy API

albertvaka opened the pull request:
### What does this PR do?

Adds a generic forwarding proxy API that can forward payloads to any intake endpoint given its path and subdomain (the domain is always set to the configured Datadog site(s)).

The proxy discards all input headers (except for `User-Agent` and `Content-Type`) and adds its own with the API key for the given site, the hostname configured in the Agent and the container tags if the request comes from a container. 

The proxy doesn't deserialize the payloads.

The code is mostly based on the [profiling proxy API](https://github.com/DataDog/datadog-agent/blob/main/pkg/trace/api/profiles.go) and the [appsec proxy API](https://github.com/DataDog/datadog-agent/blob/main/pkg/trace/appsec/appsec.go).

### Motivation

The Tracing Libraries add support for new products (profiling, appsec, ci visibility...) and need the Trace Agent to send their payloads to different endpoints. The goal of this PR is to add a generic API that can be reused for different types of payloads and endpoints instead of adding new specific forwarding APIs. The first user will be CI Visibility.

While the Tracing Libraries can send some of these payloads directly to the backend endpoints, sending them through the Agent has some advantages: you don't need to configure your Tracing Libraries with your API key or proxy settings, the hostname will match with everything else sent from the same Agent, container tags can be set...

### Additional Notes

* Naming is hard. I'm calling this EVPProxy. The api URL is `/evp_proxy` and the config settings are under `evp_proxy_conf`. I'm not sure this is the best name, but it achieves two goals: it is generic and it doesn't expose the name of internal backend components.

* Inconsistencies in header names (`Datadog-Container-ID` vs `X-Datadog-Hostname` vs `DD-API-KEY`) are to keep them consistent with other Trace Agent and backend APIs.

* Someone with access to this API (eg: a local user on a machine running the Agent) can make the Agent send a payload with the Agent API key to the current Datadog site (eg: *.datadoghq.com). This assumes this is always safe to do.

* The new configuration options are not documented. This was already the case for the profiling API settings and we can reconsider it if we see valid use cases for tweaking these settings.

### Describe how to test/QA your changes

In addition to the tests added, I tested this manually by [hacking the JS Tracing Library](https://github.com/DataDog/dd-trace-js/commit/2fc39c1b8e688872a041bb07c6b1e50ab3a726a3) so it sends CI Visibility payloads to the Agent instead of directly to the backend API.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

albertvaka committed:
[trace-agent] Initial version of the EVP Intake Proxy API

albertvaka committed:
Add metrics

albertvaka committed:
Fix broken test

albertvaka committed:
Refactor the whole thing to make it more testable

albertvaka committed:
More tests and some bug fixes

albertvaka committed:
Test all the things

albertvaka committed:
Add release note

gbbr commented on pkg/trace/config/config.go:
```diff
@@ -252,6 +252,18 @@ type ProfilingProxyConfig struct {
 }
 
+// EvpIntakeProxy ...
+type EvpIntakeProxy struct {
```
```suggestion
type EVPIntakeProxy struct {
```
We generally try to keep acronyms upper case.

gbbr commented on pkg/trace/config/config.go:
```diff
@@ -373,6 +385,9 @@ type AgentConfig struct {
 	// AppSec contains AppSec configuration.
 	AppSec AppSecConfig
 
+	// DebuggerProxy contains the settings for the Live Debugger proxy.
```
```suggestion
	// EVPIntakeProxy ...
```
Whoopsy copy 🍝 accident

gbbr commented on pkg/trace/config/config.go:
```diff
@@ -252,6 +252,18 @@ type ProfilingProxyConfig struct {
 
+// EvpIntakeProxy ...
+type EvpIntakeProxy struct {
+	// Enabled reports whether EvpIntakeProxy is enabled (defaults to apm_config.enabled).
```
You may want to double check documentations

gbbr commented on pkg/trace/api/info_test.go:
```diff
@@ -187,6 +188,7 @@ func TestInfoHandler(t *testing.T) {
 		"/v0.6/stats",
 		"/v0.1/pipeline_stats",
 		"/appsec/proxy/",
+		"/evpIntakeProxy/v1/",
```
Isn't `evp-intake-proxy` or `evp_intake_proxy` better? The former seems more friendly to general path naming style and the latter more in line with other things we already have (e.g. `pipeline_stats`).

Also, other similar paths have the `/input` suffix. Do we need that too?

gbbr commented on releasenotes/notes/trace-agent-evp-forwarding-api-000a992cbfe7c14b.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    Add support for new types of CI Visibility payloads to the Trace Agent, so
```
Not sure if this applies here but there was an unwritten rule at some point that any trace-agent change gets prefixed with `APM: `. Frankly I'm not sure it makes sense in this case.

albertvaka replied:
I didn't know that rule, but I don't think this is really "APM" anymore so I would prefer leaving it like this.

gbbr commented on pkg/trace/api/evp_intake_proxy_test.go:
```diff
@@ -0,0 +1,203 @@
+	require.Equal(t, endpoints[0].APIKey, "override_api_key")
+}
+
+func TestEvpIntakeHandler(t *testing.T) {
```
```suggestion
func TestEVPIntakeHandler(t *testing.T) {
```
Can you please capitalise "EVP" everywhere except when it is the prefix of an unexported identifier, in which case all lower-case is fine.

gbbr commented on pkg/trace/api/evp_intake_proxy_test.go:
```diff
@@ -0,0 +1,203 @@
+	conf.Site = "us3.datadoghq.com"
+	conf.Endpoints[0].APIKey = "test_api_key"
+
+	request := httptest.NewRequest("POST", "/mysubdomain/mypath/mysubpath?arg=test", nil)
```
Just `req` should do. It's the standard request notation. Mostly trying to follow https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#naming-things

gbbr commented on pkg/trace/api/evp_intake_proxy_test.go:
```diff
@@ -0,0 +1,203 @@
+	request.Header.Set("User-Agent", "test_user_agent")
+	request.Header.Set("Content-Type", "text/json")
+	request.Header.Set("Unexpected-Header", "To-Be-Discarded")
+	proxiedRequests, response, loggerOut := sendRequestThroughHandler(conf, request)
```
Same here, let's shorten them a bit. Proposing
```suggestion
	proxyreq, resp, logs := sendRequestThroughHandler(conf, request)
```

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,233 @@
+
+	// Parse request path: The first component is the target subdomain, the rest is the target path.
+	inputPath := req.URL.Path
+	urlComponents := strings.SplitN(inputPath, "/", 3)
```
```suggestion
	parts := strings.SplitN(inputPath, "/", 3)
```
We generally call this `parts`

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,233 @@
+	endpoints := evpIntakeEndpointsFromConfig(r.conf)
+	transport := r.conf.NewHTTPTransport()
+	logger := stdlog.New(log.NewThrottled(5, 10*time.Second), "EvpIntakeProxy: ", 0) // limit to 5 messages every 10 seconds
+	reverseProxyHandler := evpIntakeReverseProxyHandler(r.conf, endpoints, transport, logger)
```
I think just `handler` is enough 🙂 
```suggestion
	handler := evpIntakeReverseProxyHandler(r.conf, endpoints, transport, logger)
```

gbbr reviewed (commented):
Left a few nits, but still need to do a more thorough review.

albertvaka committed:
Better adhere to APM coding style

albertvaka committed:
Rename evp_intake_proxy/evpiproxy -> evp_proxy

albertvaka committed:
Rename EVPIntakeProxy -> EVPProxy

and evpIntakeReverseProxyHandler -> evpProxyForwarder

Zenithar commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,233 @@
+
+	var slurp *[]byte
+	if req.Body != nil {
+		body, err := ioutil.ReadAll(req.Body)
```
Consider adding an `io.LimitReader` here.

Zenithar commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,233 @@
+	}()
+
+	// Parse request path: The first component is the target subdomain, the rest is the target path.
+	inputPath := req.URL.Path
```
req and req.URL can be nil and should be tested as execution preconditions to prevent a crash if the URL is invalid via query header manipulation.

albertvaka replied:
I don't think `req` can be null: If it were, it would crash way before our code, in the reverseproxy code: https://github.com/golang/go/blob/dev.boringcrypto.go1.17/src/net/http/httputil/reverseproxy.go#L234

About `req.URL` I'm less sure, but I see no null checks on Go's own handlers like StripPrefix here: https://github.com/golang/go/blob/770e0e584a98dfd5e8d0d00558085c339fda0ed7/src/net/http/server.go#L2138

Zenithar replied:
I don't think about the fact that a pointer can't be nil or not due to parent caller nil handling. By design, a pointer can be nil and must be tested when it's coming from a different scope for a defensive posture. In go, there is no pointer reference borrowing management like in Rust. Concerning the `req.URL` to be nil, it happens when you use a custom router (like gin/echo) where the URL parsing is done externally to build the `req` instance. I had bad situations to handle due to that :-).

albertvaka replied:
Added null checks.

gbbr replied:
I disagree. By design `req` should not be `nil`. If it is ever `nil`, then that's a big programming error or an intentional malice which deserves a `panic`. We don't do this anywhere so let's not start here please.

albertvaka replied:
I also think it cannot be null (it would blow up in Go's code before reaching our code) but I didn't fully understand the case @Zenithar was talking about so I just added the checks. Happy to revise, though. And to be fair we don't do this in any other endpoint.

kayayarai reviewed (approved):
docs 👍

albertvaka committed:
split tests in subtests

albertvaka committed:
Add content-type tag to metrics

albertvaka committed:
Add tests for metrics

albertvaka committed:
Remove redundant prefixes from config fields

albertvaka committed:
Add MaxPayloadSize with 5Mb default

albertvaka committed:
Fix gofmt

albertvaka committed:
Force subdomain to be under *.evp.<site>

albertvaka committed:
Add null checks as per code review

If req is null, with the current go (1.17), we would blow up anyway in
ReverseProxy though.

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,252 @@
+		}
+	}()
+
+	if req == nil || req.URL == nil {
```
I personally don't agree with the logic in https://github.com/DataDog/datadog-agent/pull/12271/files#r892094101. `req` should not be `nil`. If it ends up being `nil`, then a panic is appropriate. Going by that logic we should check that `req` isn't `nil` everywhere. No. Please don't. I'd argue the same for `req.URL`.

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,252 @@
+		return nil, fmt.Errorf("EVPProxy: invalid request")
+	}
+
+	if req.Body != nil && t.maxPayloadSize > 0 {
```
Can `Body` ever be `nil`? I think not.

gbbr replied:
Let's leave it like that then.

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,252 @@
+	if req == nil || req.URL == nil {
+		return nil, fmt.Errorf("EVPProxy: invalid request")
+	}
+
```
With the risk of being annoying here, can you please cut down on empty lines? We're trying to follow the guidelines from our team in this repo: https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#general-guidelines (see point 3)

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,252 @@
+	// There's more than one destination endpoint
+
+	var slurp *[]byte
+	if req.Body != nil {
```
This overly defensive attitude doesn't really help us to be honest. When can `req.Body` be `nil` really? Maybe I'm unaware of some situation.

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,252 @@
+}
+
+func (t *evpProxyTransport) RoundTrip(req *http.Request) (rresp *http.Response, rerr error) {
+
```
Can you please remove this line?

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,252 @@
+
+	// There's more than one destination endpoint
+
+	var slurp *[]byte
```
Why a pointer to a slice? A slice is already a pointer by nature.

albertvaka replied:
Go n00b mistake.

albertvaka committed:
Revert "Add null checks as per code review"

This reverts commit 641f3789ff7592b14ce571b7b084b0d048ac6665.

albertvaka committed:
Add some comments

albertvaka committed:
Remove unnecessary pointer

albertvaka committed:
Revert "Force subdomain to be under *.evp.<site>"

This reverts commit 45593188df301d04c25273d6f9672a9e366aa722.

albertvaka committed:
Pass domain as a header

albertvaka committed:
Remove /input from the path

albertvaka committed:
Error out if subdomain is not set

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,234 @@
+
+	// Read the input headers
+	subdomain := req.Header.Get("X-Datadog-EVP-Subdomain")
+	containerID := req.Header.Get("Datadog-Container-ID")
```
```suggestion
	containerID := req.Header.Get(headerContainerID)
```

albertvaka replied:
Yes, I have this in mind. Depending on which one we merge first we should then update the other PR.

gbbr replied:
Thanks!

gbbr commented on pkg/trace/api/evp_intake_proxy_test.go:
```diff
@@ -0,0 +1,333 @@
+
+func TestEVPProxyForwarder(t *testing.T) {
+
+	randBodyBuf := make([]byte, 1024)
```
```suggestion
	buf := make([]byte, 1024)
```
Should be enough :)

albertvaka replied:
This is used far away from its definition, I think it's better to specify what this buffer is for.

gbbr reviewed (commented):
Few more nits.

albertvaka commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,234 @@
+		return nil, fmt.Errorf("EVPProxy: invalid subdomain: %s", subdomain)
+	}
+	metricTags = append(metricTags, "subdomain:"+subdomain)
+	if !isValidPath(req.URL.Path) {
+		return nil, fmt.Errorf("EVPProxy: invalid target path: %s", req.URL.Path)
+	}
+	if !isValidQueryString(req.URL.RawQuery) {
+		return nil, fmt.Errorf("EVPProxy: invalid query string: %s", req.URL.RawQuery)
+	}
```
We don't need to support all possible paths and domains, so we are limiting this further than what Go does. We are not accepting any valid path.

gbbr replied:
In that case, can you please document the helper functions so that it is clear how they are different from regular checks and why they are needed? Also, please put the helper functions at the end of the file, so we can start with the main structures and logic.

albertvaka commented on pkg/trace/api/evp_intake_proxy_test.go:
```diff
@@ -0,0 +1,333 @@
+	defer func(old metrics.StatsClient) { metrics.Client = old }(metrics.Client)
+	metrics.Client = stats
+
+	t.Run("happy case", func(t *testing.T) {
```
How am I supposed to express `invalid subdomain` or `max payload size reached` with a single word? Or should I just do `invalid _subdomain`?

gbbr replied:
Here are some examples for names of these tests in the order you've defined them:
* `ok`
* `containerID`
* `basic`
* `invalid-subdomain`
* `invalid-path`
* `invalid-query`
* `max`
* `override`
 
Doesn't really matter. It's an identifier ultimately.

I do notice that lately even the GO team has stopped following this style of naming tests, and have started using space-separated shorter descriptions (e.g. https://go.dev/blog/subtests), so I'll ease up on these comments moving forward 🙂

albertvaka committed:
Changes as per CR

albertvaka commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,234 @@
+		req.Body = apiutil.NewLimitedReader(req.Body, t.conf.EVPProxy.MaxPayloadSize)
+	}
+
+	// Metrics with stats for debugging
```
I opted for adding comments before each "paragraph" because this function grew quite long. I'm going to remove the most obvious ones.

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,232 @@
+		req.Body = apiutil.NewLimitedReader(req.Body, t.conf.EVPProxy.MaxPayloadSize)
+	}
+
+	beginTime := time.Now()
+	metricTags := []string{}
```
In other places in the repo we've been calling these `now` (or `start`) and `tags`. Do you mind changing that so we can be consistent? If you've seen it done otherwise in some part of the code, then sorry about that. Do let me know.

albertvaka committed:
Changes as per code review

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,232 @@
+// evpProxyHandler returns an HTTP handler for the /evp_proxy API.
+// Depending on the config, this is a proxying handler or a noop handler.
+func (r *HTTPReceiver) evpProxyHandler() http.Handler {
+	// r.conf is populated by cmd/trace-agent/config/config.go
```
you forgot this one.

gbbr commented on pkg/trace/api/evp_intake_proxy.go:
```diff
@@ -0,0 +1,232 @@
+	if !r.conf.EVPProxy.Enabled {
+		return evpProxyErrorHandler("Has been disabled in config")
+	}
+	endpoints := evpProxyEndpointsFromConfig(r.conf)
```
In theory, this could also be inside `evpProxyForwarder`

gbbr reviewed (approved):
This looks good to me. I do still have a few comments, I'll let you approach them as you will and feel free to merge after.

albertvaka committed:
More changes as per CR

albertvaka commented:
We don't want to add the new settings to the config template.

gh123man reviewed (approved):
Config changes look good to me. 
lets make sure we track and capture this on the [dual shipping docs](https://docs.datadoghq.com/agent/guide/dual-shipping/?tab=metricsservicechecks#overview) once this is GA. 

We would probably add a new tab to the doc with the new top-level `evp_proxy_config` config and an example of how to configure additional endpoints (should mostly be copy paste like the others)
